ipywidgets/static/widgets/js/services-shim.js

```diff
@@ -0,0 +1,221 @@
+     */
+    Object.defineProperty(Comm.prototype, 'comm_id', {
+        get: function() {
+            return this.jsServicesComm.commId;
+        }
+    });
+    
+    /**
+     * Target name
+     * @return {string}
+     */
+    Object.defineProperty(Comm.prototype, 'target_name', {
+        get: function() {
+            return this.jsServicesComm.targetName;
+        }
+    });
+    
+    /**
+     * Opens a sibling comm in the backend
+     * @param  {object} data
+     * @param  {object} callbacks
+     * @param  {object} metadata
+     * @return {string} msg id
+     */
+    Comm.prototype.open = function (data, callbacks, metadata) {
+        var future = this.jsServicesComm.open(data, metadata);
+        this._hookupCallbacks(future, callbacks);
+        return future.msgId;
+    };
+    
+    /**
+     * Sends a message to the sibling comm in the backend
+     * @param  {object} data
+     * @param  {object} callbacks
+     * @param  {object} metadata
+     * @param  {(ArrayBuffer | ArrayBufferView)[]} buffers
+     * @return {string} msg id
+     */
+    Comm.prototype.send = function (data, callbacks, metadata, buffers) {
+        var future = this.jsServicesComm.send(data, metadata, buffers);
+        this._hookupCallbacks(future, callbacks);
+        return future.msgId;
+    };
+    
+    /**
+     * Closes the sibling comm in the backend
+     * @param  {object} data
+     * @param  {object} callbacks
+     * @param  {object} metadata
+     * @return {string} msg id
+     */
+    Comm.prototype.close = function (data, callbacks, metadata) {
+        var future = this.jsServicesComm.close(data, metadata);
+        this._hookupCallbacks(future, callbacks);
+        return future.msgId;
+    };
+    
+    /**
+     * Register a message handler
+     * @param  {(object) => void} callback - signature of f(msg)
+     */
+    Comm.prototype.on_msg = function (callback) {
+        this.jsServicesComm.onMsg = function(data) {
+            callback.call(this, {content : {data: data}});
+        };
+    };
+    
+    /**
+     * Register a handler for when the comm is closed by the backend
+     * @param  {(object) => void} callback - signature of f(msg)
+     */
+    Comm.prototype.on_close = function (callback) {
+        this.jsServicesComm.onClose = function(data) {
+            callback.call(this, {content : {data: data}});
+        };
+    };
+    
+    /**
+     * Hooks callback object up with jupyter-js-services IKernelFuture
+     * @param  {IKernelFuture} future - jupyter-js-services IKernelFuture instance
+     * @param  {object} callbacks
+     */
+    Comm.prototype._hookupCallbacks = function(future, callbacks) {
+        if (callbacks) {
+            future.onReply = function(msg) {
+                if (callbacks.shell && callbacks.shell.reply) callbacks.shell.reply(msg);
+                // TODO: Handle payloads.  See https://github.com/jupyter/notebook/blob/master/notebook/static/services/kernels/kernel.js#L923-L947
```

Comment:
Honestly, I don't think so.  As you know, the channels are `shell`, `iopub`, and `stdin`.  The payload handler acts on shell messages.  In the spirit of minimal message processing at the kernel level and the 'hand the full message to the handler and allow it to decide what to do with it' thinking, it makes sense to *not* handle this in the kernel.